cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/TypeResolver.kt

```diff
@@ -26,40 +26,120 @@
+import de.fraunhofer.aisec.cpg.graph.types.DeclaresType
 import de.fraunhofer.aisec.cpg.graph.types.ObjectType
 import de.fraunhofer.aisec.cpg.graph.types.Type
-import de.fraunhofer.aisec.cpg.processing.IVisitor
-import de.fraunhofer.aisec.cpg.processing.strategy.Strategy
+import de.fraunhofer.aisec.cpg.graph.types.recordDeclaration
+import de.fraunhofer.aisec.cpg.helpers.SubgraphWalker
+import de.fraunhofer.aisec.cpg.passes.configuration.DependsOn
 
 /**
  * The purpose of this [Pass] is to establish a relationship between [Type] nodes (more specifically
  * [ObjectType]s) and their [RecordDeclaration].
  */
+@DependsOn(ImportResolver::class)
 open class TypeResolver(ctx: TranslationContext) : ComponentPass(ctx) {
+
+    lateinit var walker: SubgraphWalker.ScopedWalker
+
     override fun accept(component: Component) {
-        component.accept(
-            Strategy::AST_FORWARD,
-            object : IVisitor<Node>() {
-                /**
-                 * Creates the [ObjectType.recordDeclaration] relationship between [ObjectType]s and
-                 * [RecordDeclaration] with the same [Node.name].
-                 */
-                fun visit(record: RecordDeclaration) {
-                    for (t in typeManager.firstOrderTypes) {
-                        if (t.name == record.name && t is ObjectType) {
-                            // The node is the class of the type t
-                            t.recordDeclaration = record
-                        }
-                    }
+        resolveFirstOrderTypes()
+        refreshNames()
+
+        walker = SubgraphWalker.ScopedWalker(scopeManager)
+        walker.registerHandler(::handleNode)
+        walker.iterate(component)
+    }
+
+    private fun refreshNames() {
+        for (type in typeManager.secondOrderTypes) {
+            type.refreshNames()
+        }
+    }
+
+    companion object {
+        context(ContextProvider)
+        fun resolveType(type: Type): Boolean {
+            // Let's start by looking up the type according to their name and scope. We exclusively
+            // filter for nodes that implement DeclaresType, because otherwise we will get a lot of
+            // constructor declarations and such with the same name. It seems this is ok since most
+            // languages will prefer structs/classes over functions when resolving types.
+            var symbols =
+                ctx?.scopeManager?.findSymbols(type.name, startScope = type.scope) {
+                    it is DeclaresType
+                } ?: listOf()
+
+            // We need to have a single match, otherwise we have an ambiguous type and we cannot
+            // normalize it.
+            // TODO: Maybe we should have a warning in this case?
+            var declares = symbols.filterIsInstance<DeclaresType>().singleOrNull()
+
+            // Check for a possible typedef
+            var target = ctx?.scopeManager?.typedefFor(type.name, type.scope)
```

Comment:
Unfortunately, we can build loops with type alias:
```c++
namespace ABC {
    struct A {};
}

namespace ABC {
    using A = ABC::A;
}
```
We should check the resolved type alias to be not the same as the original type.
Or should the `scopeManager` take care to not return the same type?

Reply:
I think I encountered this  problem before and I had a solution like a) In mind. I will check that. thanks

Reply:
The loop should be fixed